Casting error in xview.hpp

I'm having trouble building a project of mine after upgrading xtensor and xtensor-python.  I think I've reduced it to the following test program, which compiles successfully (with clang) when I use `xtensor-python 0.19.1` (`xtensor 0.16.4`), but fails to compile when I upgrade to `xtensor-python 0.20.0` (`xtensor 0.17.4`).

If you can tell me what I'm doing wrong, I'd greatly appreciate it!

```c++
#include "xtensor/xarray.hpp"
#include "xtensor/xview.hpp"
#include "xtensor/xeval.hpp"

#define FORCE_IMPORT_ARRAY
#include "xtensor-python/pyarray.hpp"
    
int main()
{
    xt::pyarray<int> arr = xt::zeros<int>({10});
    auto v = xt::view(arr, xt::all());
    xt::eval(v);
    
    return 0;
}
```

The error is: 

```
/miniforge/envs/test-xtensor/include/xtensor/xview.hpp:1419:13: error: non-constant-expression cannot be narrowed from type 'unsigned long' to
      'value_type' (aka 'long') in initializer list [-Wc++11-narrowing]
  ...(static_cast<std::ptrdiff_t>(xt::step_size(std::get<integral_skip<S...>(I)>(m_slices), 1)) * m_e.strides()[integral_skip<S...>(I) - newaxis_count_before<S...>(integral_skip<S...>(I))])...
     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

<details>

<summary>Click for compiler version and full error</summary>

```
$ clang --version
Apple LLVM version 8.1.0 (clang-802.0.42)
Target: x86_64-apple-darwin16.7.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin

$ clang++ -std=c++14 -I${CONDA_PREFIX}/include -I ${CONDA_PREFIX}/include/python3.6m -I${CONDA_PREFIX}/lib/python3.6/site-packages/numpy/core/include -L${CONDA_PREFIX}/lib -lpython3.6m -Wl,-rpath,${CONDA_PREFIX}/lib test.cpp
In file included from test.cpp:2:
/miniforge/envs/test-xtensor/include/xtensor/xview.hpp:1419:13: error: non-constant-expression cannot be narrowed from type 'unsigned long' to
      'value_type' (aka 'long') in initializer list [-Wc++11-narrowing]
  ...(static_cast<std::ptrdiff_t>(xt::step_size(std::get<integral_skip<S...>(I)>(m_slices), 1)) * m_e.strides()[integral_skip<S...>(I) - newaxis_count_before<S...>(integral_skip<S...>(I))])...
     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/miniforge/envs/test-xtensor/include/xtensor/xview.hpp:1431:30: note: in instantiation of function template specialization
      'xt::xview<xt::pyarray<int, xt::layout_type::dynamic> &, xt::xall<unsigned long> >::compute_strides_impl<0>' requested here
        auto slice_strides = compute_strides_impl(std::make_index_sequence<n_strides>());
                             ^
/miniforge/envs/test-xtensor/include/xtensor/xview.hpp:1197:13: note: in instantiation of member function 'xt::xview<xt::pyarray<int,
      xt::layout_type::dynamic> &, xt::xall<unsigned long> >::compute_strides' requested here
            compute_strides(std::integral_constant<bool, is_contiguous_view>{});
            ^
/miniforge/envs/test-xtensor/include/xtensor/xview.hpp:1319:46: note: in instantiation of function template specialization
      'xt::xview<xt::pyarray<int, xt::layout_type::dynamic> &, xt::xall<unsigned long> >::strides<xt::pyarray<int, xt::layout_type::dynamic> >'
      requested here
            return str.size() == self(this)->strides().size() &&
                                             ^
/miniforge/envs/test-xtensor/include/xtensor/xassign.hpp:215:63: note: in instantiation of function template specialization
      'xt::xview<xt::pyarray<int, xt::layout_type::dynamic> &, xt::xall<unsigned long> >::is_trivial_broadcast<xt::svector<long, 4,
      std::__1::allocator<long>, true> >' requested here
                   (e1.layout() != layout_type::dynamic && e2.is_trivial_broadcast(e1.strides()));
                                                              ^
/miniforge/envs/test-xtensor/include/xtensor/xassign.hpp:316:53: note: in instantiation of function template specialization
      'xt::detail::is_trivial_broadcast<xt::xarray_container<xt::uvector<int, std::__1::allocator<int> >, xt::layout_type::row_major,
      xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag>, xt::xview<xt::pyarray<int,
      xt::layout_type::dynamic> &, xt::xall<unsigned long> > >' requested here
        bool trivial_broadcast = trivial && detail::is_trivial_broadcast(de1, de2);
                                                    ^
/miniforge/envs/test-xtensor/include/xtensor/xassign.hpp:339:20: note: (skipping 2 contexts in backtrace; use -ftemplate-backtrace-limit=0 to see
      all)
        base_type::assign_data(e1, e2, trivial_broadcast);
                   ^
/miniforge/envs/test-xtensor/include/xtensor/xsemantic.hpp:587:13: note: in instantiation of function template specialization
      'xt::assign_xexpression<xt::xarray_container<xt::uvector<int, std::__1::allocator<int> >, xt::layout_type::row_major, xt::svector<unsigned
      long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag>, xt::xview<xt::pyarray<int, xt::layout_type::dynamic> &,
      xt::xall<unsigned long> > >' requested here
        xt::assign_xexpression(*this, e);
            ^
/miniforge/envs/test-xtensor/include/xtensor/xsemantic.hpp:453:37: note: in instantiation of function template specialization
      'xt::xcontainer_semantic<xt::xarray_container<xt::uvector<int, std::__1::allocator<int> >, xt::layout_type::row_major, xt::svector<unsigned
      long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> >::assign_xexpression<xt::xview<xt::pyarray<int,
      xt::layout_type::dynamic> &, xt::xall<unsigned long> > >' requested here
        return this->derived_cast().assign_xexpression(e);
                                    ^
/miniforge/envs/test-xtensor/include/xtensor/xarray.hpp:439:24: note: in instantiation of function template specialization
      'xt::xsemantic_base<xt::xarray_container<xt::uvector<int, std::__1::allocator<int> >, xt::layout_type::row_major, xt::svector<unsigned long,
      4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> >::assign<xt::xview<xt::pyarray<int, xt::layout_type::dynamic> &,
      xt::xall<unsigned long> > >' requested here
        semantic_base::assign(e);
                       ^
/miniforge/envs/test-xtensor/include/xtensor/xeval.hpp:52:16: note: in instantiation of function template specialization
      'xt::xarray_container<xt::uvector<int, std::__1::allocator<int> >, xt::layout_type::row_major, xt::svector<unsigned long, 4,
      std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag>::xarray_container<xt::xview<xt::pyarray<int,
      xt::layout_type::dynamic> &, xt::xall<unsigned long> > >' requested here
        return xarray<typename I::value_type>(std::forward<T>(t));
               ^
test.cpp:12:9: note: in instantiation of function template specialization 'xt::eval<xt::xview<xt::pyarray<int, xt::layout_type::dynamic> &,
      xt::xall<unsigned long> > &, xt::xview<xt::pyarray<int, xt::layout_type::dynamic> &, xt::xall<unsigned long> > >' requested here
    xt::eval(v);
        ^
/miniforge/envs/test-xtensor/include/xtensor/xview.hpp:1419:13: note: insert an explicit cast to silence this issue
  ...(static_cast<std::ptrdiff_t>(xt::step_size(std::get<integral_skip<S...>(I)>(m_slices), 1)) * m_e.strides()[integral_skip<S...>(I) - newaxis_count_before<S...>(integral_skip<S...>(I))])...
     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     static_cast<value_type>(                                                                                                                                                                )
/miniforge/envs/test-xtensor/include/xtensor/xview.hpp:1240:13: error: non-constant-expression cannot be narrowed from type 'unsigned long' to
      'value_type' (aka 'long') in initializer list [-Wc++11-narrowing]
            (static_cast<ptrdiff_t>(xt::value(std::get<I>(m_slices), 0)) * m_e.strides()[I - newaxis_count_before<S...>(I)])...
            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/miniforge/envs/test-xtensor/include/xtensor/xview.hpp:1445:25: note: in instantiation of function template specialization
      'xt::xview<xt::pyarray<int, xt::layout_type::dynamic> &, xt::xall<unsigned long> >::data_offset_impl<0>' requested here
        m_data_offset = data_offset_impl(std::make_index_sequence<sizeof...(S)>());
                        ^
/miniforge/envs/test-xtensor/include/xtensor/xview.hpp:1197:13: note: in instantiation of member function 'xt::xview<xt::pyarray<int,
      xt::layout_type::dynamic> &, xt::xall<unsigned long> >::compute_strides' requested here
            compute_strides(std::integral_constant<bool, is_contiguous_view>{});
            ^
/miniforge/envs/test-xtensor/include/xtensor/xview.hpp:1319:46: note: in instantiation of function template specialization
      'xt::xview<xt::pyarray<int, xt::layout_type::dynamic> &, xt::xall<unsigned long> >::strides<xt::pyarray<int, xt::layout_type::dynamic> >'
      requested here
            return str.size() == self(this)->strides().size() &&
                                             ^
/miniforge/envs/test-xtensor/include/xtensor/xassign.hpp:215:63: note: in instantiation of function template specialization
      'xt::xview<xt::pyarray<int, xt::layout_type::dynamic> &, xt::xall<unsigned long> >::is_trivial_broadcast<xt::svector<long, 4,
      std::__1::allocator<long>, true> >' requested here
                   (e1.layout() != layout_type::dynamic && e2.is_trivial_broadcast(e1.strides()));
                                                              ^
/miniforge/envs/test-xtensor/include/xtensor/xassign.hpp:316:53: note: in instantiation of function template specialization
      'xt::detail::is_trivial_broadcast<xt::xarray_container<xt::uvector<int, std::__1::allocator<int> >, xt::layout_type::row_major,
      xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag>, xt::xview<xt::pyarray<int,
      xt::layout_type::dynamic> &, xt::xall<unsigned long> > >' requested here
        bool trivial_broadcast = trivial && detail::is_trivial_broadcast(de1, de2);
                                                    ^
/miniforge/envs/test-xtensor/include/xtensor/xassign.hpp:339:20: note: (skipping 2 contexts in backtrace; use -ftemplate-backtrace-limit=0 to see
      all)
        base_type::assign_data(e1, e2, trivial_broadcast);
                   ^
/miniforge/envs/test-xtensor/include/xtensor/xsemantic.hpp:587:13: note: in instantiation of function template specialization
      'xt::assign_xexpression<xt::xarray_container<xt::uvector<int, std::__1::allocator<int> >, xt::layout_type::row_major, xt::svector<unsigned
      long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag>, xt::xview<xt::pyarray<int, xt::layout_type::dynamic> &,
      xt::xall<unsigned long> > >' requested here
        xt::assign_xexpression(*this, e);
            ^
/miniforge/envs/test-xtensor/include/xtensor/xsemantic.hpp:453:37: note: in instantiation of function template specialization
      'xt::xcontainer_semantic<xt::xarray_container<xt::uvector<int, std::__1::allocator<int> >, xt::layout_type::row_major, xt::svector<unsigned
      long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> >::assign_xexpression<xt::xview<xt::pyarray<int,
      xt::layout_type::dynamic> &, xt::xall<unsigned long> > >' requested here
        return this->derived_cast().assign_xexpression(e);
                                    ^
/miniforge/envs/test-xtensor/include/xtensor/xarray.hpp:439:24: note: in instantiation of function template specialization
      'xt::xsemantic_base<xt::xarray_container<xt::uvector<int, std::__1::allocator<int> >, xt::layout_type::row_major, xt::svector<unsigned long,
      4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> >::assign<xt::xview<xt::pyarray<int, xt::layout_type::dynamic> &,
      xt::xall<unsigned long> > >' requested here
        semantic_base::assign(e);
                       ^
/miniforge/envs/test-xtensor/include/xtensor/xeval.hpp:52:16: note: in instantiation of function template specialization
      'xt::xarray_container<xt::uvector<int, std::__1::allocator<int> >, xt::layout_type::row_major, xt::svector<unsigned long, 4,
      std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag>::xarray_container<xt::xview<xt::pyarray<int,
      xt::layout_type::dynamic> &, xt::xall<unsigned long> > >' requested here
        return xarray<typename I::value_type>(std::forward<T>(t));
               ^
test.cpp:12:9: note: in instantiation of function template specialization 'xt::eval<xt::xview<xt::pyarray<int, xt::layout_type::dynamic> &,
      xt::xall<unsigned long> > &, xt::xview<xt::pyarray<int, xt::layout_type::dynamic> &, xt::xall<unsigned long> > >' requested here
    xt::eval(v);
        ^
/miniforge/envs/test-xtensor/include/xtensor/xview.hpp:1240:13: note: insert an explicit cast to silence this issue
            (static_cast<ptrdiff_t>(xt::value(std::get<I>(m_slices), 0)) * m_e.strides()[I - newaxis_count_before<S...>(I)])...
            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            static_cast<value_type>(                                                                                        )
2 errors generated.
```

</details>

--- a/include/xtensor-python/pyarray.hpp
+++ b/include/xtensor-python/pyarray.hpp
@@ -328,6 +328,7 @@ namespace xt
         using const_reference = typename base_type::const_reference;
         using pointer = typename base_type::pointer;
         using size_type = typename base_type::size_type;
+        using difference_type = typename base_type::difference_type;
         using shape_type = typename base_type::shape_type;
         using strides_type = typename base_type::strides_type;
         using backstrides_type = typename base_type::backstrides_type;
@@ -745,7 +746,7 @@ namespace xt
     {
         m_shape = inner_shape_type(reinterpret_cast<size_type*>(PyArray_SHAPE(this->python_array())),
                                    static_cast<size_type>(PyArray_NDIM(this->python_array())));
-        m_strides = inner_strides_type(reinterpret_cast<size_type*>(PyArray_STRIDES(this->python_array())),
+        m_strides = inner_strides_type(reinterpret_cast<difference_type*>(PyArray_STRIDES(this->python_array())),
                                        static_cast<size_type>(PyArray_NDIM(this->python_array())));
 
         if (L != layout_type::dynamic && !do_strides_match(m_shape, m_strides, L))

--- a/include/xtensor-python/pystrides_adaptor.hpp
+++ b/include/xtensor-python/pystrides_adaptor.hpp
@@ -27,7 +27,7 @@ namespace xt
     {
     public:
 
-        using value_type = std::size_t;
+        using value_type = std::ptrdiff_t;
         using const_reference = value_type;
         using const_pointer = const value_type*;
         using size_type = std::size_t;

--- a/test/test_pyarray.cpp
+++ b/test/test_pyarray.cpp
@@ -11,6 +11,7 @@
 #include "xtensor-python/pyarray.hpp"
 
 #include "xtensor/xarray.hpp"
+#include "xtensor/xview.hpp"
 
 #include "test_common.hpp"
 
@@ -240,5 +241,12 @@ namespace xt
         std::vector<std::size_t> sc2 = {6};
         EXPECT_TRUE(std::equal(sc2.begin(), sc2.end(), a.shape().begin()) && a.shape().size() == 1);
         EXPECT_EQ(ptr, a.data());
+    }
+
+    TEST(pyarray, view)
+    {
+        xt::pyarray<int> arr = xt::zeros<int>({ 10 });
+        auto v = xt::view(arr, xt::all());
+        EXPECT_EQ(v(0), 0.);
     }
 }

--- a/test/test_pytensor.cpp
+++ b/test/test_pytensor.cpp
@@ -11,6 +11,7 @@
 #include "xtensor-python/pytensor.hpp"
 
 #include "xtensor/xtensor.hpp"
+#include "xtensor/xview.hpp"
 
 #include "test_common.hpp"
 
@@ -230,5 +231,12 @@ namespace xt
         // note this throws because std array has only 1 element initialized
         // and the second element is `0`.
         EXPECT_THROW(a.reshape({6, 5}), std::runtime_error);
+    }
+
+    TEST(pytensor, view)
+    {
+        xt::pytensor<int, 1> arr = xt::zeros<int>({ 10 });
+        auto v = xt::view(arr, xt::all());
+        EXPECT_EQ(v(0), 0.);
     }
 }

